### ISSUE TITLE
[Feature] Support AnimateDiff

### ISSUE THREAD

**Issue by TDS4874** [Mikubill/sd-webui-controlnet#1954]
[animate-diff-support](https://github.com/TDS4874/sd-webui-controlnet/tree/animate-diff-support/tree/animate-diff-support)

I've made modifications to control AnimateDiff using ControlNet. For a overview of the trick, please check my [Twitter page](https://twitter.com/TDS_95514874) and the modifications in [scripts/hook.py](https://github.com/TDS4874/sd-webui-controlnet/blob/animate-diff-support/scripts/hook.py).

I'd like to gather opinions on whether this should be proposed as a pull request. Please share your thoughts!

https://github.com/Mikubill/sd-webui-controlnet/assets/119688242/a8485627-24af-4cdc-b959-5930dd6d36f0

**Comment by huchenlei** [comment]
lllyasviel is working on a major update to support SDXL right now, and there will be major change in `hook.py`. https://github.com/Mikubill/sd-webui-controlnet/pull/1952

I would recommend extract your logic into a separate file and wait for the above PR to land first.

**Comment by haofanwang** [comment]
Interesting! Could you make a brief explanation of your workflow so that it is more clear to those using diffusers framework? @TDS4874

**Comment by TDS4874** [comment]
Thank you for discussion. I'll wait the update.

The modifications I made to hook.py mainly fall into the following three points:

1. Defined a global variable, ANIMATEDIFF_STATE, and toggle it on and off based on whether VanillaTemporalModule exists in Unet at the start of the hook.

2. Within the `forward` function, in accordance with the order of the control mode, just before being added to `total_controlnet_embedding`, Attenuate the entire tensor based on its value in the first dimension (which corresponds to the frame in ANIMATEDIFF). The first one control near the start frame, and the second one control near the end frame.

3. Within the `hacked_basic_transformer_inner_forward` function, set both the c-side and uc-side of `self_attn1_c` to be non-applied to `self.bank`. As it gets farther from the start and end frames, it approaches from `self_attn1_uc` to `self_attn1_c`. (Note: `self_attn1_uc` is originally a tensor where both c-side and uc-side are applied to `self.bank`.)

**Comment by toyxyz** [comment]
I have been able to get satisfactory results using the openpose controlnet image sequence. Is it possible to add animatediff batch function to controlnet extension? The animatediff repo I used is this : https://github.com/s9roll7/animatediff-cli-prompt-travel

https://github.com/Mikubill/sd-webui-controlnet/assets/8006000/14fa43a9-44f8-41e7-81cf-3b23d80aecdf

https://github.com/Mikubill/sd-webui-controlnet/assets/8006000/2c375022-1985-4ae4-b1cc-1806312e0022

**Comment by tydia** [comment]
any updates on this?

**Comment by rkfg** [comment]
> 1. Defined a global variable, ANIMATEDIFF_STATE, and toggle it on and off based on whether VanillaTemporalModule exists in Unet at the start of the hook.

It's much better to put it into a field of UnetHook, it's available in `forward` via the `outer` variable

> 2. Within the `forward` function, in accordance with the order of the control mode, just before being added to `total_controlnet_embedding`, Attenuate the entire tensor based on its value in the first dimension (which corresponds to the frame in ANIMATEDIFF). The first one control near the start frame, and the second one control near the end frame.

As I understand it, the first and last elements are not attenuated for some reason, I suppose it's intentional. Can you briefly explain why?

> 3. Within the `hacked_basic_transformer_inner_forward` function, set both the c-side and uc-side of `self_attn1_c` to be non-applied to `self.bank`. As it gets farther from the start and end frames, it approaches from `self_attn1_uc` to `self_attn1_c`. (Note: `self_attn1_uc` is originally a tensor where both c-side and uc-side are applied to `self.bank`.)

I added some debug print and for me this whole branch that starts with `if len(self.bank) > 0:` doesn't even execute, at least when used with ControlNet Tile, I only need it and haven't tested with other types, so I omitted that part but if it's important and I'm doing something wrong it can explain some irregularities I described below.

So in the end I used your patch and changed it slightly to make it more robust and not use global variables:
```diff
diff --git a/scripts/hook.py b/scripts/hook.py
index 4776af4..d348f54 100644
--- a/scripts/hook.py
+++ b/scripts/hook.py
@@ -381,6 +381,7 @@ class UnetHook(nn.Module):
         self.gn_auto_machine_weight = 1.0
         self.current_style_fidelity = 0.0
         self.current_uc_indices = None
+        self.enable_animate_diff = False
 
     @staticmethod
     def call_vae_using_process(p, x, batch_size=None, mask=None):
@@ -558,7 +559,7 @@ class UnetHook(nn.Module):
                 context = torch.cat([context, control.clone()], dim=1)
 
             # handle ControlNet / T2I_Adapter
-            for param in outer.control_params:
+            for param_idx, param in enumerate(outer.control_params):
                 if no_high_res_control:
                     continue
 
@@ -629,7 +630,28 @@ class UnetHook(nn.Module):
                 if param.global_average_pooling:
                     control = [torch.mean(c, dim=(2, 3), keepdim=True) for c in control]
 
+                frfactors = [0.5, 0.3, 0.2, 0.1]
+                factor_count = len(frfactors)
                 for idx, item in enumerate(control):
+                    frame_count = round(len(item) / 2)
+                    if outer.enable_animate_diff and frame_count > factor_count+3:
+                        for ifr, frfactor in enumerate(frfactors):
+                            if param_idx == 0:
+                                item[ifr+1] *= frfactor
+                                item[frame_count+ifr+1] *= frfactor
+
+                            if param_idx == 1:
+                                item[frame_count*2-ifr-2] *= frfactor
+                                item[frame_count-ifr-2] *= frfactor
+
+                        if param_idx == 0:
+                            item[factor_count+1:frame_count] = 0
+                            item[frame_count+factor_count+1:] = 0
+
+                        if param_idx == 1:
+                            item[:frame_count-factor_count-1] = 0
+                            item[frame_count:frame_count*2-factor_count-1] = 0
+
                     target = None
                     if param.control_model_type == ControlModelType.ControlNet:
                         target = total_controlnet_embedding
@@ -943,6 +965,10 @@ class UnetHook(nn.Module):
 
         all_modules = torch_dfs(model)
 
+        for i, module in enumerate(all_modules):
+            if "VanillaTemporalModule" in module.__class__.__name__:
+                self.enable_animate_diff = True
+
         if need_attention_hijack:
             attn_modules = [module for module in all_modules if isinstance(module, BasicTransformerBlock) or isinstance(module, BasicTransformerBlockSGM)]
             attn_modules = sorted(attn_modules, key=lambda x: - x.norm1.normalized_shape[0])
```
Thank you, it's a great method! There's also a way to *end* the animation with an image but it's not very intuitive. If I simply disable the 0th unit the 1st one is used as the starting image, there's no apparent way to find the actual unit number inside the extension. So instead I keep them both on but set the weight of the 0th unit to zero, it shouldn't affect the generation in this case.

I noticed that quite often the animation doesn't continue the first image and instead quickly fade into something else. Usually simply trying another seed helps. Perhaps more controlling methods are needed or this one can be improved to keep the model in the context.

**Comment by oliverban** [comment]
This is so cool! :) Any updates on it?

**Comment by TDS4874** [comment]
@rkfg 
> It's much better to put it into a field of UnetHook, it's available in `forward` via the `outer` variable

I agree with it.

>As I understand it, the first and last elements are not attenuated for some reason, I suppose it's intentional. Can you briefly explain why?

That's the concept. It's to generate between two images. The two images remain unchanged.
Of course, if the overall weight is reduced on the UI side, the weight of the first and last frames will also decrease.

>So in the end I used your patch and changed it slightly to make it more robust and not use global variables:

I also think it would be good to do it like param_idx, but it won't work correctly if the user inserts something other than the Control type into the second unit. It's troubling...

>I added some debug print and for me this whole branch that starts with if len(self.bank) > 0: doesn't even execute, at least when used with ControlNet Tile, I only need it and haven't tested with other types, so I omitted that part but if it's important and I'm doing something wrong it can explain some irregularities I described below.

As for 'Reference only', it was working in my tests, but perhaps it's not robust. A deeper discussion is needed regarding 'Reference only'. We might be able to skip it for now.
In 'Reference only', the original forward function is used to generate attention. However, this original already has the motion module from AnimateDiff injected. This causes a phenomenon where the diversity between frames is lost, and the video becomes static.
I tested skipping the motion module during attention generation, and the quality of the video improves while retaining the influence of the Reference. However, this means that we also need to make changes to the AnimateDiff repository.

**Comment by lllyasviel** [comment]
will take a look soon when fixing some other things in XL controls

**Comment by rkfg** [comment]
@TDS4874 thank you for the answers, yes, makes sense! I thought the first and last item in that tensor are somewhat special hence the confusion.

I haven't tried the reference mode but most of the others seem to not use that transformer function. I agree that it needs discussion and some sort of good UI. For now I made these frame factors customizable in UI in my fork: https://github.com/rkfg/sd-webui-controlnet/tree/animatediff

![2023-09-20_01-02-48](https://github.com/Mikubill/sd-webui-controlnet/assets/184066/453b2a6a-25f7-44c6-b6f5-c117256be9c8)

**Comment by akk-123** [comment]
Any updates on it? @lllyasviel
